src/resourceIndex.js

```diff
@@ -0,0 +1,129 @@
+   * (i.e. no versions)
+   * @returns {LATEST} if the versionDescriptor is more recent than the most
+   * recent version in the index.
+   * @returns {String} the YYYY-MM-DD version ID
+   */
+  versionIdFromDescriptor(versionDescriptor) {
+    if (!versionDescriptor.match(/^\d{4}-\d{2}-\d{2}$/)) {
+      throw new BadRequest(`Version descriptor ("${versionDescriptor}" must be YYYY-MM-DD`)
+    }
+
+    const dates = (this.data?.versions || [])
+      .map((v) => v.date)
+      .sort();
+
+    if (!dates.length) {
+      throw new NotFound(`Attempted to lookup a version for ${versionDescriptor} however this resource has no versions in the index`)
+    }
+    if (versionDescriptor < dates[0]) {
+      throw new NotFound(`Version descriptor ${versionDescriptor} predates available versions`)
+    }
+    if (versionDescriptor > dates[dates.length - 1]) {
+      return LATEST;
+    }
+
+    let id;
+    for (const date of dates) {
+      if (date <= versionDescriptor) {
+        id = date;
+      }
+    }
+    return id;
+  }
+
+  /**
+   * Given a YYYY-MM-DD string, return the available subresources and their corresponding URLs
+   */
+  subresourceUrls(date) {
+    for (const version of this.data?.versions || []) {
+      if (version.date===date) {
+        return version.fileUrls;
+      }
+    }
+    throw new Error(`Attempted to access a specific date which doesn't exist in the resource index: ${date}`)
+  }
+
+}
+
+
+/**
+ * Updates provided data in-place by fetching the file on disk.
+ * We expose an update function rather than automatically updating here. This
+ * prevents scripts / tests which (indirectly) import this file unintentionally
+ * fetching stuff from S3 (and often failing due to missing credentials).
+ */
+async function updateResourceVersions() {
+  if (RESOURCE_INDEX) {
+    utils.verbose(`Updating available resources index from ${RESOURCE_INDEX}`);
+    let fileContents = fs.readFileSync(RESOURCE_INDEX)
+    if (RESOURCE_INDEX.endsWith('.gz')) {
+      fileContents = await gunzip(fileContents)
+    }
+    resources = JSON.parse(fileContents);
+  } else {
+    const [Bucket, Key] = ['nextstrain-inventories', 'resources.json.gz'];
+    utils.verbose(`Updating available resources index from s3://${Bucket}/${Key}`);
+    const S3 = new AWS.S3();
+    try {
+      const newETag = (await S3.headObject({Bucket, Key}).promise()).ETag;
+      if (newETag && newETag === eTag) {
+        utils.verbose("Skipping available resource update as eTag hasn't changed");
+        return;
+      }
+      /* NOTE: in AWS SDK v2 response.Body is a buffer; in v3 it's a stream */
+      const newResources = JSON.parse(
+        await S3.getObject({Bucket, Key}).promise()
+          .then((response) => gunzip(response.Body))
+      );
+      [resources, eTag] = [newResources, newETag];
+    } catch (err) {
+      utils.warn(`Resource updating failed: ${err.message}`)
+      return;
+    }
```

Comment:
This is the code I was thinking of when [I said](https://github.com/nextstrain/nextstrain.org/pull/719#issuecomment-1789356224):

> Adopt AWS config support added to master since this was branched. Probably swap away from v2 SDK and add necessary functions to src/s3.js?

Consolidating around our own S3 access layer helps keep its configuration centralized and standardized (e.g. `src/aws.js`) and expose a more ergonomic API for our uses (`src/s3.js`). (The former was necessary as part of AWS configurability for CDC.)

Reply:
I guess this is the added complexity of allowing others to deploy nextstrain.org. Can you take a look at [0b56c84](https://github.com/nextstrain/nextstrain.org/pull/719/commits/0b56c84c9c30fafda817b18bfd6842b45ba40097)?

Reply:
I guess centralized configuration can be "added complexity" from some viewpoints, but it's also "removed complexity" from other viewpoints...

I looked at and commented on that commit in my most recent review, so I'll resolve this comment now.